### PR TITLE
Task specific vocab, word to sequence and gro2env random goal

### PR DESCRIPTION
**Pull request by Jialn**
Changes:
* Task specific vocab
* Teacher merges the vocab from every tasks
* Sentence to sequence helper function to teacher.py
* Random goal from the teacher's task for gro2env

Sentence to sequence is implemented by looking up to a dict, and sequence to sentence is by looking up to the sorted vocab list. Thus both are quite fast.

### CODE REVIEW

File: python/social_bot/teacher.py [hunk 1]
```diff
@@ -145,6 +146,25 @@ def add_task_group(self, task_group, weight=1):
         self._task_groups.append(task_group)
         self._weights.append(weight)
 
+    def build_vocab_from_tasks(self):
+        vovab_list = []
```
Comment (emailweixu) [inline python/social_bot/teacher.py]: vovab => vocab here and other places

File: python/social_bot/teacher_tasks.py [hunk 2]
```diff
@@ -56,6 +56,7 @@ def __init__(self,
         self._fail_distance_thresh = fail_distance_thresh
         self._max_steps = max_steps
         self._random_range = random_range
+        self.task_vocab = ['hello', 'goal', 'Well', 'done!', 'Failed', 'to']
```
Comment (emailweixu) [inline python/social_bot/teacher_tasks.py]: 'done!' => 'done'
And let's use lower cases for all words right now.

File: python/social_bot/teacher.py [hunk 3]
```diff
@@ -145,6 +146,25 @@ def add_task_group(self, task_group, weight=1):
...
+            for t in g._tasks:
+                vovab_list = vovab_list + t.task_vocab
+        # remove repeated words and convert to dict
+        self._vovab_list = sorted(set(vovab_list),key=vovab_list.index)
+        self.vocab_size = len(self._vovab_list)
+        self._vovab_dict = dict(zip(self._vovab_list,list(range(0,self.vocab_size))))
+
+    def sentence_to_sequence(self, sentence):
+        word_list = sentence.split()
+        sequence = list(map(lambda x : self._vovab_dict[x], word_list))
```
Comment (emailweixu) [inline python/social_bot/teacher.py]: report error if word cannot be found.

File: python/social_bot/teacher.py [hunk 4]
```diff
@@ -145,6 +146,25 @@ def add_task_group(self, task_group, weight=1):
...
+        return np.array(sequence)
```
Comment (emailweixu) [inline python/social_bot/teacher.py]: length of sequence needs to match the length given in observation space.  We can fill the remaining values using 0 (other word ID start from 1). And we need to check if sequence exceeds the length limit.

File: python/social_bot/teacher.py [hunk 5]
```diff
@@ -145,6 +148,32 @@ def add_task_group(self, task_group, weight=1):
...
         self._weights.append(weight)
 
+    def build_vocab_from_tasks(self):
+        vocab_list = []
+        for g in self._task_groups:
+            for t in g._tasks:
+                vocab_list = vocab_list + t.task_vocab
+        # remove repeated words and convert to dict
+        self._vocab_list = sorted(set(vocab_list),key=vocab_list.index)
+        self.vocab_size = len(self._vocab_list)
```
Comment (emailweixu) [inline python/social_bot/teacher.py]: need to +1 to include the "0" for padding.

File: python/social_bot/teacher.py [hunk 6]
```diff
@@ -145,6 +148,32 @@ def add_task_group(self, task_group, weight=1):
...
+        self._vocab_dict = dict(
+            zip(self._vocab_list, list(range(1, 1 + self.vocab_size))))
+
+    def sentence_to_sequence(self, sentence, max_sequence_length):
+        word_list = sentence.split()
+        sequence = list(map(lambda x : self._vocab_dict[x], word_list))
+        if None in sequence:
+            logger.warn("Word out of vocab:" + word_list[sequence.index(None)])
+        padding_num = max_sequence_length - len(sequence)
+        assert padding_num > 0
```
Comment (emailweixu) [inline python/social_bot/teacher.py]: padding_num >= 0, "some error message for how to fix"

File: python/social_bot/teacher.py [hunk 7]
```diff
@@ -145,6 +148,32 @@ def add_task_group(self, task_group, weight=1):
...
+                vocab_list = vocab_list + t.task_vocab
+        # remove repeated words and convert to dict
+        self._vocab_list = sorted(set(vocab_list),key=vocab_list.index)
+        self.vocab_size = len(self._vocab_list)
+        self._vocab_dict = dict(
+            zip(self._vocab_list, list(range(1, 1 + self.vocab_size))))
+
+    def sentence_to_sequence(self, sentence, max_sequence_length):
+        word_list = sentence.split()
+        sequence = list(map(lambda x : self._vocab_dict[x], word_list))
```
Comment (emailweixu) [inline python/social_bot/teacher.py]: if x cannot be found, this line will crash

File: python/social_bot/teacher.py [hunk 8]
```diff
@@ -145,6 +148,32 @@ def add_task_group(self, task_group, weight=1):
...
+        if None in sequence:
+            logger.warn("Word out of vocab:" + word_list[sequence.index(None)])
+        padding_num = max_sequence_length - len(sequence)
+        assert padding_num > 0
+        return np.pad(sequence, (0, padding_num), 'constant')
+
+    def sequence_to_sentence(self, sequence):
+        for seq_len in range(len(sequence)):
+            if sequence[seq_len] == 0: break
+        word_list = list(map(lambda x : self._vocab_list[x-1], sequence[:seq_len]))
```
Comment (emailweixu) [inline python/social_bot/teacher.py]: Check if x is out of range.

File: python/social_bot/teacher.py [hunk 9]
```diff
@@ -145,6 +148,32 @@ def add_task_group(self, task_group, weight=1):
         self._task_groups.append(task_group)
         self._weights.append(weight)
 
+    def build_vocab_from_tasks(self):
+        vocab_list = []
+        for g in self._task_groups:
+            for t in g._tasks:
+                vocab_list = vocab_list + t.task_vocab
+        # remove repeated words and convert to dict
+        self._vocab_list = sorted(set(vocab_list),key=vocab_list.index)
```
Comment (emailweixu) [inline python/social_bot/teacher.py]: space  after ","

File: python/social_bot/teacher.py (hunk 5 above)
Comment (Jialn) [reply python/social_bot/teacher.py]: operation '+1' is done when building the dict
`        self._vocab_dict = dict(
            zip(self._vocab_list, list(range(1, 1 + self.vocab_size))))`

Comment (Jialn) [reply python/social_bot/teacher.py]: Yes, we should.

**Pull request merged**